File: linkmint-backend/ledger-go/ledger.go
package ledger

import (
	"context"
	"fmt"
	"math/big"
	"strings"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
)

// DBTX is the subset of pgx used by the ledger helpers. *pgxpool.Pool, *pgx.Conn, and pgx.Tx all
// satisfy it, so a caller can post ledger legs on its OWN transaction — the business-state write and
// the ledger write then commit together (A.6) or roll back together. The helpers never Begin/Commit;
// the caller owns the unit of work.
type DBTX interface {
	Exec(ctx context.Context, sql string, args ...any) (pgconn.CommandTag, error)
	Query(ctx context.Context, sql string, args ...any) (pgx.Rows, error)
	QueryRow(ctx context.Context, sql string, args ...any) pgx.Row
}

// selectColumns casts uuid/numeric to text so values scan cleanly into Go types regardless of pgx
// codec defaults (the repo convention — see audit-log-service). amount::text round-trips the full
// NUMERIC(38,0) into *big.Int without a numeric codec.
const selectColumns = `id, entry_group::text, account, direction, amount::text, currency, pl_id, note, created_at`

// Post validates a balanced set of legs and writes them atomically under one entry_group. The legs
// are inserted in a single statement (atomic), so a group is all-or-nothing even on a pool; when db
// is a caller's pgx.Tx the legs join that transaction. Returns the entry_group. Rejects unbalanced
// or malformed input (ErrUnbalanced / ErrInvalidLeg) before touching the database.
func Post(ctx context.Context, db DBTX, in PostingInput) (uuid.UUID, error) {
	if err := validate(in.Entries); err != nil {
		return uuid.Nil, err
	}

	group := uuid.New()
	if in.EntryGroup != nil {
		group = *in.EntryGroup
	}
	plID := nullIfEmpty(in.PLID)
	note := nullIfEmpty(in.Note)

	var b strings.Builder
	b.WriteString(`INSERT INTO ledger.ledger_entries (entry_group, account, direction, amount, currency, pl_id, note) VALUES `)
	args := make([]any, 0, len(in.Entries)*7)
	for i, leg := range in.Entries {
		if i > 0 {
			b.WriteByte(',')
		}
		n := i * 7
		fmt.Fprintf(&b, "($%d::uuid,$%d,$%d,$%d::numeric,$%d,$%d,$%d)", n+1, n+2, n+3, n+4, n+5, n+6, n+7)
		args = append(args, group.String(), leg.Account, string(leg.Direction), leg.Amount.String(), leg.Currency, plID, note)
	}

	if _, err := db.Exec(ctx, b.String(), args...); err != nil {
		return uuid.Nil, fmt.Errorf("ledger post: %w", err)
	}
	return group, nil
}

// Reverse posts a correcting entry group: it reads the original group's legs and writes a NEW group
// with every direction flipped (DR↔CR), same amounts/currency/account/pl_id. The original is never
// edited or deleted (A.6). Returns the new entry_group. note defaults to "reversal of <group>".
func Reverse(ctx context.Context, db DBTX, group uuid.UUID, note string) (uuid.UUID, error) {
	entries, err := EntriesByGroup(ctx, db, group)
	if err != nil {
		return uuid.Nil, err
	}
	if len(entries) == 0 {
		return uuid.Nil, fmt.Errorf("%w: %s", ErrGroupNotFound, group)
	}

	flipped := make([]Leg, len(entries))
	plID := ""
	for i, e := range entries {
		flipped[i] = Leg{Account: e.Account, Direction: flip(e.Direction), Amount: e.Amount, Currency: e.Currency}
		if plID == "" && e.PLID != "" {
			// First non-empty pl_id wins — matches the Python reverse() (next(...)). A group's legs
			// all share one pl_id, so this only ever differs for externally-corrupted data.
			plID = e.PLID
		}
	}
	if note == "" {
		note = "reversal of " + group.String()
	}
	return Post(ctx, db, PostingInput{Entries: flipped, PLID: plID, Note: note})
}

// EntriesByGroup returns all legs of an entry_group, oldest id first.
func EntriesByGroup(ctx context.Context, db DBTX, group uuid.UUID) ([]Entry, error) {
	rows, err := db.Query(ctx,
		`SELECT `+selectColumns+` FROM ledger.ledger_entries WHERE entry_group=$1::uuid ORDER BY id`,
		group.String())
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// EntriesByAccount returns the most recent legs for an account (newest first), capped by limit
// (default 100, max 1000). For reporting/reconciliation reads.
func EntriesByAccount(ctx context.Context, db DBTX, account string, limit int) ([]Entry, error) {
	rows, err := db.Query(ctx,
		`SELECT `+selectColumns+` FROM ledger.ledger_entries WHERE account=$1 ORDER BY created_at DESC, id DESC LIMIT $2`,
		account, clampLimit(limit))
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// EntriesByPLID returns the most recent legs tied to a pl_id (newest first), capped by limit.
func EntriesByPLID(ctx context.Context, db DBTX, plID string, limit int) ([]Entry, error) {
	rows, err := db.Query(ctx,
		`SELECT `+selectColumns+` FROM ledger.ledger_entries WHERE pl_id=$1 ORDER BY created_at DESC, id DESC LIMIT $2`,
		plID, clampLimit(limit))
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// Balance returns the net balance of an account in a currency as ΣCR − ΣDR (the credit-positive
// convention). A read-only aggregate — it never moves funds (A.1).
func Balance(ctx context.Context, db DBTX, account, currency string) (*big.Int, error) {
	var s string
	if err := db.QueryRow(ctx,
		`SELECT COALESCE(SUM(CASE WHEN direction='CR' THEN amount ELSE -amount END), 0)::text
		   FROM ledger.ledger_entries WHERE account=$1 AND currency=$2`,
		account, currency).Scan(&s); err != nil {
		return nil, err
	}
	v, ok := new(big.Int).SetString(s, 10)
	if !ok {
		return nil, fmt.Errorf("ledger: unparseable balance %q for %s/%s", s, account, currency)
	}
	return v, nil
}

// IsBalanced reports whether the whole ledger nets to zero for a currency (ΣDR == ΣCR across every
// account). The global double-entry invariant work27 reconciliation checks.
func IsBalanced(ctx context.Context, db DBTX, currency string) (bool, error) {
	var s string
	if err := db.QueryRow(ctx,
		`SELECT COALESCE(SUM(CASE WHEN direction='DR' THEN amount ELSE -amount END), 0)::text
		   FROM ledger.ledger_entries WHERE currency=$1`,
		currency).Scan(&s); err != nil {
		return false, err
	}
	return s == "0", nil
}

func clampLimit(n int) int {
	switch {
	case n <= 0:
		return 100
	case n > 1000:
		return 1000
	default:
		return n
	}
}

func nullIfEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func scanEntries(rows pgx.Rows) ([]Entry, error) {
	var out []Entry
	for rows.Next() {
		var (
			e       Entry
			groupS  string
			dir     string
			amountS string
			plID    *string
			note    *string
		)
		if err := rows.Scan(&e.ID, &groupS, &e.Account, &dir, &amountS, &e.Currency, &plID, &note, &e.CreatedAt); err != nil {
			return nil, err
		}
		g, err := uuid.Parse(groupS)
		if err != nil {
			return nil, fmt.Errorf("ledger: unparseable entry_group %q: %w", groupS, err)
		}
		amt, ok := new(big.Int).SetString(amountS, 10)
		if !ok {
			return nil, fmt.Errorf("ledger: unparseable amount %q for entry %d", amountS, e.ID)
		}
		e.EntryGroup = g
		e.Direction = Direction(dir)
		e.Amount = amt
		if plID != nil {
			e.PLID = *plID
		}
		if note != nil {
			e.Note = *note
		}
		out = append(out, e)
	}
	return out, rows.Err()
}

File: linkmint-backend/ledger-python/src/linkmint_ledger/ledger.py
"""Async posting/reversal/read helpers over a caller's SQLAlchemy connection.

Mirrors ledger-go/ledger.go. The helpers run on the caller's ``AsyncConnection`` or
``AsyncSession`` and NEVER commit — the caller's unit-of-work owns the transaction boundary, so a
business-state write and the ledger legs commit together (A.6) or roll back together. Non-custodial
(A.1): this records flows between opaque account labels (e.g. ``paylink:PLK...``, ``treasury``,
``validator:0x...``); it never holds or moves funds. Amounts are exact integers (minor units);
balances are read-only sums.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncConnection, AsyncSession

from .entry import Direction, Leg, _flip, validate
from .errors import GroupNotFound

# A caller may pass either an AsyncConnection or an AsyncSession; both share one transaction.
Conn = AsyncConnection | AsyncSession

_SELECT = (
    "SELECT id, entry_group::text AS entry_group, account, direction, amount::text AS amount, "
    "currency, pl_id, note, created_at FROM ledger.ledger_entries"
)


@dataclass(frozen=True)
class Entry:
    """A persisted ledger row (one leg) as read back from the table."""

    id: int
    entry_group: uuid.UUID
    account: str
    direction: Direction
    amount: int
    currency: str
    pl_id: str | None
    note: str | None
    created_at: datetime


async def _as_conn(db: Conn) -> AsyncConnection:
    """Resolve an AsyncSession to its underlying AsyncConnection; pass a connection through."""
    if isinstance(db, AsyncSession):
        return await db.connection()
    return db


def _clamp(limit: int) -> int:
    if limit <= 0:
        return 100
    if limit > 1000:
        return 1000
    return limit


def _row_to_entry(row: object) -> Entry:
    return Entry(
        id=row.id,  # type: ignore[attr-defined]
        entry_group=uuid.UUID(row.entry_group),  # type: ignore[attr-defined]
        account=row.account,  # type: ignore[attr-defined]
        direction=Direction(row.direction),  # type: ignore[attr-defined]
        amount=int(row.amount),  # type: ignore[attr-defined]
        currency=row.currency,  # type: ignore[attr-defined]
        pl_id=row.pl_id,  # type: ignore[attr-defined]
        note=row.note,  # type: ignore[attr-defined]
        created_at=row.created_at,  # type: ignore[attr-defined]
    )


async def post(
    db: Conn,
    *,
    entries: list[Leg],
    entry_group: uuid.UUID | None = None,
    pl_id: str | None = None,
    note: str | None = None,
) -> uuid.UUID:
    """Validate a balanced set of legs and write them atomically under one entry_group.

    The legs are inserted in a single statement (atomic), so a group is all-or-nothing; when ``db``
    is a caller's session/connection inside a transaction, the legs join it. Returns the
    entry_group. Rejects unbalanced/malformed input (:class:`Unbalanced` / :class:`InvalidLeg`)
    before any write.
    """
    validate(entries)
    group = entry_group or uuid.uuid4()
    conn = await _as_conn(db)

    rows: list[str] = []
    params: dict[str, object] = {"g": str(group), "pl": pl_id, "note": note}
    for i, leg in enumerate(entries):
        # CAST(... AS ...) rather than ::uuid / ::numeric — SQLAlchemy text() will not bind a
        # ``:name`` that is immediately followed by ``::``. Amounts bind as strings (exact NUMERIC).
        rows.append(
            f"(CAST(:g AS uuid), :acct{i}, :dir{i}, "
            f"CAST(:amt{i} AS numeric), :ccy{i}, :pl, :note)"
        )
        params[f"acct{i}"] = leg.account
        params[f"dir{i}"] = Direction(leg.direction).value
        params[f"amt{i}"] = str(leg.amount)
        params[f"ccy{i}"] = leg.currency

    sql = (
        "INSERT INTO ledger.ledger_entries "
        "(entry_group, account, direction, amount, currency, pl_id, note) VALUES " + ", ".join(rows)
    )
    await conn.execute(text(sql), params)
    return group


async def reverse(db: Conn, entry_group: uuid.UUID, *, note: str | None = None) -> uuid.UUID:
    """Post a correcting entry group: read the original group's legs and write a NEW group with
    every direction flipped (DR<->CR), same amounts/currency/account/pl_id. The original is never
    edited or deleted (A.6). Returns the new entry_group. Raises :class:`GroupNotFound` if empty.
    """
    entries = await entries_by_group(db, entry_group)
    if not entries:
        raise GroupNotFound(str(entry_group))
    pl_id = next((e.pl_id for e in entries if e.pl_id), None)
    flipped = [
        Leg(account=e.account, direction=_flip(e.direction), amount=e.amount, currency=e.currency)
        for e in entries
    ]
    return await post(db, entries=flipped, pl_id=pl_id, note=note or f"reversal of {entry_group}")


async def entries_by_group(db: Conn, entry_group: uuid.UUID) -> list[Entry]:
    """Return all legs of an entry_group, oldest id first."""
    conn = await _as_conn(db)
    result = await conn.execute(
        text(_SELECT + " WHERE entry_group = CAST(:g AS uuid) ORDER BY id"),
        {"g": str(entry_group)},
    )
    return [_row_to_entry(r) for r in result.all()]


async def entries_by_account(db: Conn, account: str, limit: int = 100) -> list[Entry]:
    """Return the most recent legs for an account (newest first), capped by ``limit`` (max 1000)."""
    conn = await _as_conn(db)
    result = await conn.execute(
        text(_SELECT + " WHERE account=:a ORDER BY created_at DESC, id DESC LIMIT :n"),
        {"a": account, "n": _clamp(limit)},
    )
    return [_row_to_entry(r) for r in result.all()]


async def entries_by_pl_id(db: Conn, pl_id: str, limit: int = 100) -> list[Entry]:
    """Most recent legs tied to a pl_id (newest first), capped by ``limit`` (max 1000)."""
    conn = await _as_conn(db)
    result = await conn.execute(
        text(_SELECT + " WHERE pl_id=:p ORDER BY created_at DESC, id DESC LIMIT :n"),
        {"p": pl_id, "n": _clamp(limit)},
    )
    return [_row_to_entry(r) for r in result.all()]


async def balance(db: Conn, account: str, currency: str) -> int:
    """Net balance of an account in a currency as ΣCR − ΣDR (credit-positive). Read-only (A.1)."""
    conn = await _as_conn(db)
    result = await conn.execute(
        text(
            "SELECT COALESCE(SUM(CASE WHEN direction='CR' THEN amount ELSE -amount END), 0)::text "
            "FROM ledger.ledger_entries WHERE account=:a AND currency=:c"
        ),
        {"a": account, "c": currency},
    )
    return int(result.scalar_one())


async def is_balanced(db: Conn, currency: str) -> bool:
    """Whether the whole ledger nets to zero for a currency (ΣDR == ΣCR). work27 reconciliation."""
    conn = await _as_conn(db)
    result = await conn.execute(
        text(
            "SELECT COALESCE(SUM(CASE WHEN direction='DR' THEN amount ELSE -amount END), 0)::text "
            "FROM ledger.ledger_entries WHERE currency=:c"
        ),
        {"c": currency},
    )
    return int(result.scalar_one()) == 0
